src/tools/r2rdump/R2RMethod.cs

```diff
@@ -0,0 +1,267 @@
+
+            int val = 0;
+            int baseOffset = DecodeUnsigned(image, methodDefEntryPointsOffset, ref val);
+            int nElements = (int)(val >> 2);
+            int entryIndexSize = (int)(val & 3);
+            int offset = MethodDefEntryPointsTryGetAt(image, rid - 1, (int)baseOffset, nElements, entryIndexSize);
+            int id = 0;
+            offset = DecodeUnsigned(image, offset, ref id);
+            if ((id & 1)!=0)
+            {
+                if ((id & 2)!= 0)
+                {
+                    val = 0;
+                    DecodeUnsigned(image, offset, ref val);
+                    offset -= val;
+                }
+                id >>= 2;
+            }
+            else
+            {
+                id >>= 1;
+            }
+            NativeCode = runtimeFunctions[id];
+        }
+
+        private int MethodDefEntryPointsTryGetAt(byte[] image, int index, int baseOffset, int nElements, int entryIndexSize)
+        {
+            if (index >= nElements)
+                throw new System.BadImageFormatException("MethodDefEntryPoints");
+
+            int offset = 0;
+            if (entryIndexSize == 0)
+            {
+                int i = baseOffset + (index / _blockSize);
+                offset = image[i];
+            }
+            else if (entryIndexSize == 1)
+            {
+                int i = baseOffset + 2 * (index / _blockSize);
+                offset = R2RReader.ReadUInt16(image, ref i);
+            }
+            else
+            {
+                int i = baseOffset + 4 * (index / _blockSize);
+                offset = R2RReader.ReadInt32(image, ref i);
+            }
+            offset += baseOffset;
+
+            for (uint bit = _blockSize >> 1; bit > 0; bit >>= 1)
+            {
+                int val = 0;
+                int offset2 = DecodeUnsigned(image, offset, ref val);
+                if ((index & bit) != 0)
+                {
+                    if ((val & 2) != 0)
+                    {
+                        offset = offset + (val >> 2);
+                        continue;
+                    }
+                }
+                else
+                {
+                    if ((val & 1) != 0)
+                    {
+                        offset = offset2;
+                        continue;
+                    }
+                }
+
+                // Not found
+                if ((val & 3) == 0)
+                {
+                    // Matching special leaf node?
+                    if ((val >> 2) == (index & (_blockSize - 1)))
+                    {
+                        offset = offset2;
+                        break;
+                    }
+                }
+                throw new System.BadImageFormatException("MethodDefEntryPoints");
+            }
+            return offset;
+        }
+
+        private int DecodeUnsigned(byte[] image, int offset, ref int pValue)
+        {
+            int val = image[offset];
```

Comment:
val should be a uint for the DecodeUnsigned method

Reply:
In the future, I think it would be good to bounds checking in this method, and throw an exception -- probably BadImageFormatException.